[Question] TS compiler errors in @theia/workspace package

I have an npm package which depends on `@theia/languages` which pulls in `@theia/workspace` dependency.
When I attempt to TS compile my package (with TS 3.1.3, I've tried 2.8.x) i get these errors:

```
ode_modules/@theia/filesystem/lib/browser/file-tree/file-tree-model.d.ts:48:45 - error TS2304: Cannot find name 'WebKitEntry'.

48     protected uploadEntry(base: URI, entry: WebKitEntry | null): void;
                                               ~~~~~~~~~~~

node_modules/@theia/filesystem/lib/browser/file-tree/file-tree-model.d.ts:49:54 - error TS2304: Cannot find name 'WebKitDirectoryEntry'.

49     protected uploadDirectoryEntry(base: URI, entry: WebKitDirectoryEntry): Promise<void>;
                                                        ~~~~~~~~~~~~~~~~~~~~

node_modules/@theia/filesystem/lib/browser/file-tree/file-tree-model.d.ts:54:34 - error TS2304: Cannot find name 'WebKitDirectoryEntry'.

54     protected readEntries(entry: WebKitDirectoryEntry, cb: (items: any) => void): void;
                                    ~~~~~~~~~~~~~~~~~~~~

node_modules/@theia/filesystem/lib/browser/file-tree/file-tree-model.d.ts:55:49 - error TS2304: Cannot find name 'WebKitEntry'.

55     protected uploadEntries(base: URI, entries: WebKitEntry[]): void;
                                                   ~~~~~~~~~~~

node_modules/@theia/filesystem/lib/browser/file-tree/file-tree-model.d.ts:56:49 - error TS2304: Cannot find name 'WebKitFileEntry'.

56     protected uploadFileEntry(base: URI, entry: WebKitFileEntry): void;
                                                   ~~~~~~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:135:28 - error TS1254: A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.

135     const validateSchema = new Ajv().compile({
                               ~~~~~~~~~~~~~~~~~~~
136         type: 'object',
    ~~~~~~~~~~~~~~~~~~~~~~~
... 
151         }
    ~~~~~~~~~
152     });
    ~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:135:32 - error TS2304: Cannot find name 'Ajv'.

135     const validateSchema = new Ajv().compile({
                                   ~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:153:58 - error TS1183: An implementation cannot be declared in ambient contexts.

153     export function is(data: any): data is WorkspaceData {
                                                             ~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:156:74 - error TS1183: An implementation cannot be declared in ambient contexts.

156     export function buildWorkspaceData(folders: string[]): WorkspaceData {
                                                                             ~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:161:103 - error TS1183: An implementation cannot be declared in ambient contexts.

161     export function transformToRelative(data: WorkspaceData, workspaceFile?: FileStat): WorkspaceData {
                                                                                                          ~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:162:38 - error TS1039: Initializers are not allowed in ambient contexts.

162         const folderUris: string[] = [];
                                         ~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:163:34 - error TS1254: A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.

163         const workspaceFileUri = new URI(workspaceFile ? workspaceFile.uri : '').withScheme('file');
                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:165:31 - error TS1254: A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.

165             const folderUri = new URI(path).withScheme('file');
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:166:25 - error TS1254: A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.

166             const rel = workspaceFileUri.parent.relative(folderUri);
                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:167:13 - error TS1036: Statements are not allowed in ambient contexts.

167             if (rel) {
                ~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:168:17 - error TS1036: Statements are not allowed in ambient contexts.

168                 folderUris.push(rel.toString());
                    ~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:171:17 - error TS1036: Statements are not allowed in ambient contexts.

171                 folderUris.push(folderUri.toString());
                    ~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:176:103 - error TS1183: An implementation cannot be declared in ambient contexts.

176     export function transformToAbsolute(data: WorkspaceData, workspaceFile?: FileStat): WorkspaceData {
                                                                                                          ~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:178:39 - error TS1039: Initializers are not allowed in ambient contexts.

178             const folders: string[] = [];
                                          ~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:179:13 - error TS1036: Statements are not allowed in ambient contexts.

179             for (const folder of data.folders) {
                ~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:180:30 - error TS1254: A 'const' initializer in an ambient context must be a string or numeric literal or literal enum reference.

180                 const path = folder.path;
                                 ~~~~~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:181:17 - error TS1036: Statements are not allowed in ambient contexts.

181                 if (path.startsWith('file:///')) {
                    ~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:182:21 - error TS1036: Statements are not allowed in ambient contexts.

182                     folders.push(path);
                        ~~~~~~~

node_modules/@theia/workspace/lib/browser/workspace-service.d.ts:185:21 - error TS1036: Statements are not allowed in ambient contexts.

185                     folders.push(new URI(workspaceFile.uri).withScheme('file').parent.resolve(path).toString());
                        ~~~~~~~

```

They are coming mostly from `@theia/workspace` and its dependencies.

Any idea how to fix or workaround that?

If I try typescript 2.7.2 exactly i get these TS compiler errors instead:
```
node_modules/@theia/core/lib/common/types.d.ts(17,5): error TS1131: Property or signature expected.
node_modules/@theia/core/lib/common/types.d.ts(17,27): error TS1005: ']' expected.
node_modules/@theia/core/lib/common/types.d.ts(17,28): error TS1005: ';' expected.
node_modules/@theia/core/lib/common/types.d.ts(17,29): error TS1128: Declaration or statement expected.
node_modules/@theia/core/lib/common/types.d.ts(18,1): error TS1128: Declaration or statement expected.
node_modules/@types/prop-types/index.d.ts(29,46): error TS1005: ';' expected.
node_modules/@types/prop-types/index.d.ts(29,75): error TS1005: ';' expected.
node_modules/@types/prop-types/index.d.ts(29,99): error TS1005: ';' expected.
node_modules/@types/prop-types/index.d.ts(31,54): error TS1005: ';' expected.
node_modules/@types/prop-types/index.d.ts(31,78): error TS1005: ';' expected.
node_modules/@types/prop-types/index.d.ts(31,81): error TS1109: Expression expected.
node_modules/@types/prop-types/index.d.ts(31,97): error TS1109: Expression expected.
node_modules/@types/prop-types/index.d.ts(31,122): error TS1005: ';' expected.
node_modules/@types/prop-types/index.d.ts(31,130): error TS1128: Declaration or statement expected.
node_modules/@types/prop-types/index.d.ts(31,138): error TS1005: ',' expected.
node_modules/@types/prop-types/index.d.ts(44,48): error TS1005: ';' expected.
node_modules/@types/prop-types/index.d.ts(44,49): error TS1109: Expression expected.
node_modules/@types/prop-types/index.d.ts(44,50): error TS1109: Expression expected.
node_modules/@types/prop-types/index.d.ts(44,68): error TS1005: '(' expected.
node_modules/@types/prop-types/index.d.ts(44,69): error TS1005: ')' expected.
node_modules/@types/prop-types/index.d.ts(46,30): error TS1005: ';' expected.
node_modules/@types/prop-types/index.d.ts(46,54): error TS1005: ';' expected.
node_modules/@types/prop-types/index.d.ts(46,57): error TS1109: Expression expected.
node_modules/@types/react/index.d.ts(2297,27): error TS1005: ';' expected.
node_modules/@types/react/index.d.ts(2298,14): error TS1005: ':' expected.
node_modules/@types/react/index.d.ts(2298,28): error TS1005: ';' expected.
node_modules/@types/react/index.d.ts(2299,9): error TS1109: Expression expected.
node_modules/@types/react/index.d.ts(2300,9): error TS1005: '(' expected.
node_modules/@types/react/index.d.ts(2301,9): error TS1005: '(' expected.
node_modules/@types/react/index.d.ts(2302,5): error TS1005: '(' expected.
node_modules/@types/react/index.d.ts(2302,12): error TS1005: ')' expected.
node_modules/@types/react/index.d.ts(2314,49): error TS1005: ';' expected.
node_modules/@types/react/index.d.ts(2314,76): error TS1005: ';' expected.
node_modules/@types/react/index.d.ts(2314,99): error TS1005: ';' expected.
node_modules/@types/react/index.d.ts(2315,13): error TS1128: Declaration or statement expected.
node_modules/@types/react/index.d.ts(2316,13): error TS1005: '(' expected.
node_modules/@types/react/index.d.ts(2316,17): error TS1005: ',' expected.
node_modules/@types/react/index.d.ts(2316,44): error TS1005: ',' expected.
node_modules/@types/react/index.d.ts(2316,45): error TS1005: ':' expected.
node_modules/@types/react/index.d.ts(2318,17): error TS1005: '(' expected.
node_modules/@types/react/index.d.ts(2318,21): error TS1005: ',' expected.
node_modules/@types/react/index.d.ts(2318,51): error TS1005: ',' expected.
node_modules/@types/react/index.d.ts(2318,52): error TS1005: ':' expected.
node_modules/@types/react/index.d.ts(2320,21): error TS1005: '(' expected.
node_modules/@types/react/index.d.ts(2320,24): error TS1005: ')' expected.
```

--- a/package.json
+++ b/package.json
@@ -39,7 +39,7 @@
     "tslint": "^5.10.0",
     "tslint-language-service": "^0.9.9",
     "typedoc": "^0.8",
-    "typescript": "^2.7.2",
+    "typescript": "^3.1.3",
     "uuid": "^3.1.0",
     "wdio-mocha-framework": "0.5.9",
     "wdio-selenium-standalone-service": "0.0.8",

--- a/packages/console/src/browser/console-widget.ts
+++ b/packages/console/src/browser/console-widget.ts
@@ -139,7 +139,10 @@ export class ConsoleWidget extends BaseWidget implements StatefulWidget {
     }
 
     selectAll(): void {
-        document.getSelection().selectAllChildren(this.content.node);
+        const selection = document.getSelection();
+        if (selection) {
+            selection.selectAllChildren(this.content.node);
+        }
     }
 
     collapseAll(): void {

--- a/packages/core/src/browser/tree/search-box.ts
+++ b/packages/core/src/browser/tree/search-box.ts
@@ -15,7 +15,7 @@
  ********************************************************************************/
 
 import { SearchBoxDebounce, SearchBoxDebounceOptions } from '../tree/search-box-debounce';
-import { BaseWidget } from '../widgets/widget';
+import { BaseWidget, Message } from '../widgets/widget';
 import { Emitter, Event } from '../../common/event';
 import { KeyCode, Key } from '../keys';
 
@@ -172,7 +172,6 @@ export class SearchBox extends BaseWidget {
 
         const input = document.createElement('input');
         input.readOnly = true;
-        input.onselectstart = () => false;
         input.type = 'text';
         input.classList.add(
             SearchBox.Styles.SEARCH_INPUT,
@@ -221,6 +220,12 @@ export class SearchBox extends BaseWidget {
             close
         };
 
+    }
+
+    protected onAfterAttach(msg: Message): void {
+        super.onAfterAttach(msg);
+        // tslint:disable-next-line:no-any
+        this.addEventListener(this.input, 'selectstart' as any, () => false);
     }
 
 }

--- a/packages/core/src/browser/widgets/widget.ts
+++ b/packages/core/src/browser/widgets/widget.ts
@@ -112,7 +112,7 @@ export class BaseWidget extends Widget {
     }
 
     protected addEventListener<K extends keyof HTMLElementEventMap>(element: HTMLElement, type: K, listener: EventListenerOrEventListenerObject<K>, useCapture?: boolean): void {
-        this.toDisposeOnDetach.push(addEventListener(element, type, listener));
+        this.toDisposeOnDetach.push(addEventListener(element, type, listener, useCapture));
     }
 
     protected addKeyListener<K extends keyof HTMLElementEventMap>(

--- a/packages/debug/src/browser/console/debug-console-items.tsx
+++ b/packages/debug/src/browser/console/debug-console-items.tsx
@@ -201,8 +201,9 @@ export class DebugVariable extends ExpressionContainer {
         return !!this.valueRef && document.queryCommandSupported('copy');
     }
     copyValue(): void {
-        if (this.valueRef) {
-            document.getSelection().selectAllChildren(this.valueRef);
+        const selection = document.getSelection();
+        if (this.valueRef && selection) {
+            selection.selectAllChildren(this.valueRef);
             document.execCommand('copy');
         }
     }
@@ -213,8 +214,9 @@ export class DebugVariable extends ExpressionContainer {
         return !!this.nameRef && document.queryCommandSupported('copy');
     }
     copyAsExpression(): void {
-        if (this.nameRef) {
-            document.getSelection().selectAllChildren(this.nameRef);
+        const selection = document.getSelection();
+        if (this.nameRef && selection) {
+            selection.selectAllChildren(this.nameRef);
             document.execCommand('copy');
         }
     }

--- a/packages/debug/src/browser/debug-frontend-application-contribution.ts
+++ b/packages/debug/src/browser/debug-frontend-application-contribution.ts
@@ -601,8 +601,9 @@ export class DebugFrontendApplicationContribution extends AbstractViewContributi
         registry.registerCommand(DebugCommands.COPY_CALL_STACK, {
             execute: () => {
                 const { frames } = this;
-                if (frames) {
-                    document.getSelection().selectAllChildren(frames.node);
+                const selection = document.getSelection();
+                if (frames && selection) {
+                    selection.selectAllChildren(frames.node);
                     document.execCommand('copy');
                 }
             },

--- a/packages/filesystem/src/browser/file-tree/file-tree-model.ts
+++ b/packages/filesystem/src/browser/file-tree/file-tree-model.ts
@@ -237,11 +237,11 @@ export class FileTreeModel extends TreeModelImpl implements LocationService {
 
     protected uploadFile(base: URI, file: File): void {
         const reader = new FileReader();
-        reader.onload = () => this.uploadFileContent(base.resolve(file.name), reader.result);
+        reader.onload = () => this.uploadFileContent(base.resolve(file.name), reader.result as ArrayBuffer);
         reader.readAsArrayBuffer(file);
     }
 
-    protected async uploadFileContent(base: URI, fileContent: Iterable<number>): Promise<void> {
+    protected async uploadFileContent(base: URI, fileContent: ArrayBuffer): Promise<void> {
         const uri = base.toString();
         const encoding = 'base64';
         const content = base64.fromByteArray(new Uint8Array(fileContent));

--- a/packages/filesystem/src/typings/dom.webkit.d.ts
+++ b/packages/filesystem/src/typings/dom.webkit.d.ts
@@ -1,0 +1,75 @@
+/********************************************************************************
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+type WebKitEntriesCallback = ((entries: WebKitEntry[]) => void) | { handleEvent(entries: WebKitEntry[]): void; };
+
+type WebKitErrorCallback = ((err: DOMError) => void) | { handleEvent(err: DOMError): void; };
+
+type WebKitFileCallback = ((file: File) => void) | { handleEvent(file: File): void; };
+
+interface WebKitDirectoryEntry extends WebKitEntry {
+    createReader(): WebKitDirectoryReader;
+}
+
+declare var WebKitDirectoryEntry: {
+    prototype: WebKitDirectoryEntry;
+    new(): WebKitDirectoryEntry;
+};
+
+interface WebKitDirectoryReader {
+    readEntries(successCallback: WebKitEntriesCallback, errorCallback?: WebKitErrorCallback): void;
+}
+
+declare var WebKitDirectoryReader: {
+    prototype: WebKitDirectoryReader;
+    new(): WebKitDirectoryReader;
+};
+
+interface WebKitEntry {
+    readonly filesystem: WebKitFileSystem;
+    readonly fullPath: string;
+    readonly isDirectory: boolean;
+    readonly isFile: boolean;
+    readonly name: string;
+}
+
+declare var WebKitEntry: {
+    prototype: WebKitEntry;
+    new(): WebKitEntry;
+};
+
+interface WebKitFileEntry extends WebKitEntry {
+    file(successCallback: WebKitFileCallback, errorCallback?: WebKitErrorCallback): void;
+}
+
+declare var WebKitFileEntry: {
+    prototype: WebKitFileEntry;
+    new(): WebKitFileEntry;
+};
+
+interface WebKitFileSystem {
+    readonly name: string;
+    readonly root: WebKitDirectoryEntry;
+}
+
+declare var WebKitFileSystem: {
+    prototype: WebKitFileSystem;
+    new(): WebKitFileSystem;
+};
+
+declare interface DataTransferItem {
+    webkitGetAsEntry(): WebKitEntry | null;
+}

--- a/packages/mini-browser/src/browser/mini-browser.ts
+++ b/packages/mini-browser/src/browser/mini-browser.ts
@@ -455,7 +455,7 @@ export class MiniBrowser extends BaseWidget {
         }
         try {
             const { contentDocument } = this.frame;
-            if (contentDocument) {
+            if (contentDocument && contentDocument.location) {
                 location = contentDocument.location.href;
             }
         } catch {

--- a/packages/monaco/src/browser/monaco-quick-open-service.ts
+++ b/packages/monaco/src/browser/monaco-quick-open-service.ts
@@ -57,7 +57,7 @@ export class MonacoQuickOpenService extends QuickOpenService {
 
     internalOpen(opts: MonacoQuickOpenControllerOpts): void {
         this.opts = opts;
-        this.previousActiveElement = window.document.activeElement;
+        this.previousActiveElement = window.document.activeElement || undefined;
         this.widget.show(this.opts.prefix || '');
         this.setPlaceHolder(opts.inputAriaLabel);
         this.setPassword(opts.password ? true : false);

--- a/packages/navigator/src/browser/navigator-widget.tsx
+++ b/packages/navigator/src/browser/navigator-widget.tsx
@@ -97,16 +97,18 @@ export class FileNavigatorWidget extends FileTreeWidget {
 
     protected enableDndOnMainPanel(): void {
         const mainPanelNode = this.shell.mainPanel.node;
-        this.addEventListener(mainPanelNode, 'drop', async e => {
-            const treeNode = this.getTreeNodeFromData(e.dataTransfer);
+        this.addEventListener(mainPanelNode, 'drop', async ({ dataTransfer }) => {
+            const treeNode = dataTransfer && this.getTreeNodeFromData(dataTransfer) || undefined;
 
             if (FileNode.is(treeNode)) {
                 this.commandService.executeCommand(CommonCommands.OPEN.id, treeNode.uri);
             }
         });
         const handler = (e: DragEvent) => {
-            e.dataTransfer.dropEffect = 'link';
-            e.preventDefault();
+            if (e.dataTransfer) {
+                e.dataTransfer.dropEffect = 'link';
+                e.preventDefault();
+            }
         };
         this.addEventListener(mainPanelNode, 'dragover', handler);
         this.addEventListener(mainPanelNode, 'dragenter', handler);

--- a/packages/preview/src/browser/markdown/markdown-preview-handler.ts
+++ b/packages/preview/src/browser/markdown/markdown-preview-handler.ts
@@ -259,7 +259,9 @@ export class MarkdownPreviewHandler implements PreviewHandler {
                 const elements = body.getElementsByTagName(tag);
                 for (let i = 0; i < elements.length; i++) {
                     const element = elements.item(i);
-                    procedure(element);
+                    if (element) {
+                        procedure(element);
+                    }
                 }
             };
 

--- a/packages/terminal/src/browser/terminal-widget-impl.ts
+++ b/packages/terminal/src/browser/terminal-widget-impl.ts
@@ -183,7 +183,7 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         }
 
         /* Get the CSS properties of <html> (aka :root in css).  */
-        const htmlElementProps = getComputedStyle(document.documentElement);
+        const htmlElementProps = getComputedStyle(document.documentElement!);
 
         const fontFamily = lookup(htmlElementProps, '--theia-terminal-font-family');
         const fontSizeStr = lookup(htmlElementProps, '--theia-code-font-size');

